File: pulsaros-welcome/usr/share/pulsaros-welcome/welcome.py
#!/usr/bin/env python3
# ==============================================================================
# Pulsar OS - Welcome Application (Python + WebKitGTK Exact React Port)
# ==============================================================================

import os
import re
import sys
import json
import shutil
import subprocess
import cairo

import gi
gi.require_version("Gtk", "3.0")
gi.require_version("Gdk", "3.0")
gi.require_version("GdkPixbuf", "2.0")

# Support both WebKit2 4.1 and 6.0
try:
    gi.require_version("WebKit2", "4.1")
    from gi.repository import WebKit2
except (ValueError, ImportError):
    try:
        gi.require_version("WebKit2", "6.0")
        from gi.repository import WebKit2
    except (ValueError, ImportError):
        gi.require_version("WebKit2", "4.0")
        from gi.repository import WebKit2

from gi.repository import Gtk, Gdk, GLib, Gio


def is_live_system():
    """Detects if we are running in the Live ISO environment."""
    if os.path.exists("/lib/live/mount") or os.path.exists("/run/archiso/bootmnt") or os.path.exists("/run/live/medium"):
        return True
    if os.environ.get("USER") == "live":
        return True
    try:
        with open("/proc/cmdline", "r") as f:
            cmd = f.read()
            if "boot=live" in cmd or "archisobasedir=live" in cmd or "rootfstype=9p" in cmd:
                return True
    except Exception:
        pass
    return False


def is_arch_system():
    """Detects if running on Arch Linux."""
    try:
        with open("/etc/os-release", "r") as f:
            rel = f.read().lower()
        if "id_like=arch" in rel or "id=arch" in rel:
            return True
        if "id_like=debian" in rel or "id=debian" in rel or "id=ubuntu" in rel:
            return False
    except Exception:
        pass
    return shutil.which("pacman") is not None


def is_ootb_pending():
    """Detects if initial user account setup is required."""
    return os.path.exists("/etc/pulsar-need-setup")


def check_sentinel():
    """Returns True if the user has already completed the welcome onboarding."""
    done_file = os.path.expanduser("~/.config/pulsaros-welcome.done")
    return os.path.isfile(done_file)


def write_sentinel():
    """Marks welcome assistant as completed."""
    done_file = os.path.expanduser("~/.config/pulsaros-welcome.done")
    os.makedirs(os.path.dirname(done_file), exist_ok=True)
    try:
        with open(done_file, "w") as f:
            f.write("done\n")
    except Exception as e:
        print(f"Error writing sentinel: {e}")


def launch_with_fallback(primary, fallback=None):
    candidates = [primary]
    if fallback:
        candidates.append(fallback)
    for cmd in candidates:
        if "/" in cmd:
            if os.path.exists(cmd) and os.access(cmd, os.X_OK):
                subprocess.Popen([cmd])
                return True
        elif shutil.which(cmd):
            subprocess.Popen([cmd])
            return True
    return False


def get_system_resolutions():
    resolutions = []
    try:
        out = subprocess.check_output(["xrandr"], universal_newlines=True)
        for line in out.splitlines():
            m = re.search(r"^\s*(\d+)x(\d+)\s+.*(\*|\+)", line)
            if m:
                w, h = int(m.group(1)), int(m.group(2))
                active = "*" in line
                resolutions.append({"width": w, "height": h, "active": active})
    except Exception:
        resolutions = [{"width": 1920, "height": 1080, "active": True}]
    return resolutions


class WelcomeApp(Gtk.Window):
    def __init__(self, is_live=False, is_arch=False, needs_ootb=False):
        super().__init__(type=Gtk.WindowType.TOPLEVEL)
        self.is_live = is_live
        self.is_arch = is_arch
        self.needs_ootb = needs_ootb

        self.set_title("Pulsar OS Welcome")
        self.set_decorated(False)
        self.maximize()

        # Enable true transparent alpha channel for GNOME blur-my-shell
        screen = self.get_screen()
        visual = screen.get_rgba_visual()
        if visual and screen.is_composited():
            self.set_visual(visual)

        self.set_app_paintable(True)
        self.connect("draw", self._on_draw)
        self.connect("destroy", Gtk.main_quit)

        # Set up WebKit with bridge
        ucm = WebKit2.UserContentManager()
        ucm.register_script_message_handler("welcome")
        ucm.connect("script-message-received::welcome", self._on_script_message)

        # Inject initial runtime state
        init_script = WebKit2.UserScript(
            f"""
                window.IS_LIVE = {'true' if is_live else 'false'};
                window.IS_ARCH = {'true' if is_arch else 'false'};
                window.IS_OOTB = {'true' if needs_ootb else 'false'};
            """,
            WebKit2.UserContentInjectedFrames.ALL_FRAMES,
            WebKit2.UserScriptInjectionTime.START,
            None,
            None
        )
        ucm.add_script(init_script)

        self.webview = WebKit2.WebView.new_with_user_content_manager(ucm)
        bg_color = Gdk.RGBA()
        bg_color.parse("rgba(0, 0, 0, 0)")
        self.webview.set_background_color(bg_color)

        settings = self.webview.get_settings()
        settings.set_allow_file_access_from_file_urls(True)
        settings.set_allow_universal_access_from_file_urls(True)
        settings.set_enable_write_console_messages_to_stdout(True)
        settings.set_enable_developer_extras(True)

        self.add(self.webview)

        # Load exact compiled React/Tailwind frontend from dist/index.html (or fallback to ui/index.html)
        base_dir = os.path.dirname(os.path.abspath(__file__))
        dist_html = os.path.join(base_dir, "dist", "index.html")
        ui_html = os.path.join(base_dir, "ui", "index.html")
        
        target_html = dist_html if os.path.exists(dist_html) else ui_html
        self.webview.load_uri(f"file://{target_html}")

    def _on_draw(self, widget, cr):
        cr.set_source_rgba(0, 0, 0, 0)
        cr.set_operator(cairo.OPERATOR_SOURCE)
        cr.paint()
        return False

    def _send_response(self, req_id, result):
        if not req_id:
            return
        res_json = json.dumps(result)
        js = f"if (window.__handlePyResponse) {{ window.__handlePyResponse('{req_id}', {res_json}); }}"
        self.webview.run_javascript(js, None, None, None)

    def _on_script_message(self, content_manager, js_result):
        try:
            if hasattr(js_result, "get_js_value"):
                js_val = js_result.get_js_value()
                msg_str = js_val.to_string()
            elif hasattr(js_result, "to_string"):
                msg_str = js_result.to_string()
            else:
                msg_str = str(js_result)
            data = json.loads(msg_str)
        except Exception as e:
            print(f"Error parsing script message: {e}")
            return

        cmd = data.get("cmd") or data.get("action")
        args = data.get("args") or {}
        req_id = data.get("id")

        if cmd == "is_live_system":
            self._send_response(req_id, self.is_live)

        elif cmd == "is_arch_system":
            self._send_response(req_id, self.is_arch)

        elif cmd == "is_ootb_pending":
            self._send_response(req_id, self.needs_ootb)

        elif cmd == "check_sentinel":
            self._send_response(req_id, check_sentinel())

        elif cmd == "write_sentinel":
            write_sentinel()
            self._send_response(req_id, None)

        elif cmd == "get_system_mode":
            self._send_response(req_id, "Live" if self.is_live else "Normal")

        elif cmd == "get_resolutions":
            self._send_response(req_id, get_system_resolutions())

        elif cmd == "set_resolution":
            w = args.get("width")
            h = args.get("height")
            if w and h:
                subprocess.Popen(["xrandr", "-s", f"{w}x{h}"])
            self._send_response(req_id, None)

        elif cmd == "get_effects_state":
            self._send_response(req_id, True)

        elif cmd == "set_effects":
            self._send_response(req_id, None)

        elif cmd == "check_adb_devices":
            try:
                out = subprocess.check_output(["adb", "devices"], universal_newlines=True)
            except Exception:
                out = ""
            self._send_response(req_id, out)

        elif cmd == "launch_ootb":
            self.hide()
            candidates = ["/usr/bin/pulsaros-ootb", "/usr/share/pulsaros/welcome_ootb.py"]
            for cand in candidates:
                if os.path.exists(cand):
                    if cand.endswith(".py"):
                        subprocess.Popen(["sudo", "-E", "/usr/bin/python3", cand])
                    else:
                        subprocess.Popen([cand])
                    break
            self._send_response(req_id, None)
            Gtk.main_quit()

        elif cmd == "launch_recovery":
            self.hide()
            launch_with_fallback("pulsaros-recovery", "pulsaros-recovery-window")
            self._send_response(req_id, None)
            Gtk.main_quit()

        elif cmd == "launch_wifi_settings" or cmd == "launch_wifi":
            if not launch_with_fallback("gnome-control-center", None):
                launch_with_fallback("nm-connection-editor")
            else:
                subprocess.Popen(["gnome-control-center", "wifi"])
            self._send_response(req_id, None)

        elif cmd == "launch_bluetooth_settings" or cmd == "launch_bluetooth":
            if not launch_with_fallback("gnome-control-center", None):
                launch_with_fallback("blueman-manager")
            else:
                subprocess.Popen(["gnome-control-center", "bluetooth"])
            self._send_response(req_id, None)

        elif cmd == "launch_display_settings" or cmd == "launch_display":
            subprocess.Popen(["gnome-control-center", "display"])
            self._send_response(req_id, None)

        elif cmd == "launch_appearance_settings" or cmd == "launch_appearance":
            subprocess.Popen(["gnome-control-center", "background"])
            self._send_response(req_id, None)

        elif cmd == "launch_app":
            app = args.get("app") or data.get("app")
            fallback = args.get("fallback") or data.get("fallback")
            if app:
                launch_with_fallback(app, fallback)
            self._send_response(req_id, None)

        elif cmd == "open_url":
            url = args.get("url") or data.get("url")
            if url:
                try:
                    Gio.AppInfo.launch_default_for_uri(url, None)
                except Exception:
                    subprocess.Popen(["xdg-open", url])
            self._send_response(req_id, None)

        elif cmd == "close" or cmd == "close_window":
            self.destroy()
            Gtk.main_quit()


def main():
    force = any(arg in sys.argv for arg in ["--force", "-f"])
    is_live = is_live_system()
    is_arch = is_arch_system()
    needs_ootb = is_ootb_pending()

    # If live user cleanup is pending, launch the cleanup spinner app first
    if os.path.exists("/etc/pulsar-need-cleanup"):
        if os.path.exists("/usr/bin/pulsar-cleanup-user"):
            subprocess.Popen(["/usr/bin/pulsar-cleanup-user"])
            sys.exit(0)

    # If already done and not live/ootb/forced, exit silently
    if not is_live and not needs_ootb and check_sentinel() and not force:
        sys.exit(0)

    # Permit root X11 permissions for OOTB assistant
    if needs_ootb and shutil.which("xhost"):
        subprocess.run(["xhost", "+SI:localuser:root"], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)

    win = WelcomeApp(is_live=is_live, is_arch=is_arch, needs_ootb=needs_ootb)
    win.show_all()
    Gtk.main()


if __name__ == "__main__":
    main()

File: pulsaros-welcome/usr/share/pulsaros-welcome/src-tauri/src/welcome.rs
use std::path::Path;
use std::process::Command;

const BLUR_MY_SHELL_UUID: &str = "blur-my-shell@aunetx";
const LIQUID_GLASS_UUID: &str = "liquid-glass@thinkingcoding1231.gmail.com";

// ── System Detection ──

pub fn is_live_system() -> bool {
    if Path::new("/lib/live/mount").exists() {
        return true;
    }
    if let Ok(user) = std::env::var("USER") {
        if user == "live" {
            return true;
        }
    }
    if let Ok(cmdline) = std::fs::read_to_string("/proc/cmdline") {
        if cmdline.contains("boot=live") || cmdline.contains("rootfstype=9p") {
            return true;
        }
    }
    false
}

pub fn is_arch_system() -> bool {
    if let Ok(rel) = std::fs::read_to_string("/etc/os-release") {
        let lower = rel.to_lowercase();
        if lower.contains("id_like=arch") || lower.contains("id=arch") {
            return true;
        }
        if lower.contains("id_like=debian") || lower.contains("id=debian") || lower.contains("id=ubuntu") {
            return false;
        }
    }
    if Command::new("which").arg("pacman").output().map(|o| o.status.success()).unwrap_or(false) {
        return true;
    }
    if Command::new("which").arg("apt-get").output().map(|o| o.status.success()).unwrap_or(false) {
        return false;
    }
    false
}

// ── Sentinel ──

pub fn sentinel_path() -> String {
    let home = std::env::var("HOME").unwrap_or_else(|_| "/root".into());
    format!("{}/.config/pulsaros-welcome.done", home)
}

pub fn check_sentinel() -> bool {
    Path::new(&sentinel_path()).exists()
}

pub fn write_sentinel() -> Result<(), String> {
    let path = sentinel_path();
    if let Some(parent) = Path::new(&path).parent() {
        std::fs::create_dir_all(parent).map_err(|e| e.to_string())?;
    }
    std::fs::write(&path, "").map_err(|e| e.to_string())
}

// ── Resolution ──

#[derive(Clone, serde::Serialize)]
pub struct Resolution {
    pub width: u32,
    pub height: u32,
    pub active: bool,
}

pub fn get_resolutions() -> Vec<Resolution> {
    let output = Command::new("xrandr")
        .output()
        .map(|o| String::from_utf8_lossy(&o.stdout).to_string())
        .unwrap_or_default();

    let mut resolutions = Vec::new();
    for line in output.lines() {
        let parts: Vec<&str> = line.split_whitespace().collect();
        if parts.len() < 2 {
            continue;
        }
        if !parts[0].contains('x') {
            continue;
        }
        let dims: Vec<&str> = parts[0].split('x').collect();
        if dims.len() != 2 {
            continue;
        }
        let w: u32 = match dims[0].parse() {
            Ok(v) => v,
            Err(_) => continue,
        };
        let h: u32 = match dims[1].parse() {
            Ok(v) => v,
            Err(_) => continue,
        };
        let active = parts[1..].iter().any(|s| *s == "*");
        if !resolutions.iter().any(|r: &Resolution| r.width == w && r.height == h) {
            resolutions.push(Resolution {
                width: w,
                height: h,
                active,
            });
        }
    }
    resolutions
}

// ── App Launcher ──

pub fn launch_app_with_fallback(primary: &str, fallback: Option<&str>) -> Result<(), String> {
    if which(primary) {
        return Command::new(primary)
            .spawn()
            .map(|_| ())
            .map_err(|e| format!("Failed to launch {}: {}", primary, e));
    }
    if let Some(fb) = fallback {
        if which(fb) {
            return Command::new(fb)
                .spawn()
                .map(|_| ())
                .map_err(|e| format!("Failed to launch {}: {}", fb, e));
        }
    }
    Err(format!("{} not found in PATH", primary))
}

fn which(cmd: &str) -> bool {
    Command::new("which")
        .arg(cmd)
        .output()
        .map(|o| o.status.success())
        .unwrap_or(false)
}

// ── Display Settings ──

pub fn launch_display_settings() -> Result<(), String> {
    Command::new("gnome-control-center")
        .arg("display")
        .spawn()
        .map_err(|e| e.to_string())?;
    Ok(())
}

pub fn launch_wifi_settings() -> Result<(), String> {
    Command::new("gnome-control-center")
        .arg("wifi")
        .spawn()
        .map_err(|e| e.to_string())?;
    Ok(())
}

pub fn launch_bluetooth_settings() -> Result<(), String> {
    Command::new("gnome-control-center")
        .arg("bluetooth")
        .spawn()
        .map_err(|e| e.to_string())?;
    Ok(())
}

// ── ADB ──

pub fn check_adb_devices() -> Result<String, String> {
    let out = Command::new("adb")
        .arg("devices")
        .output()
        .map_err(|e| format!("Error running adb: {}", e))?;
    Ok(String::from_utf8_lossy(&out.stdout).to_string())
}

// ── GSettings Helpers ──

fn gsettings_get(schema: &str, key: &str) -> Result<String, String> {
    let out = Command::new("gsettings")
        .args(["get", schema, key])
        .output()
        .map_err(|e| e.to_string())?;
    Ok(String::from_utf8_lossy(&out.stdout).trim().to_string())
}

fn gsettings_set(schema: &str, key: &str, value: &str) -> Result<(), String> {
    Command::new("gsettings")
        .args(["set", schema, key, value])
        .output()
        .map_err(|e| e.to_string())?;
    Ok(())
}

fn get_enabled_extensions() -> Vec<String> {
    let raw = gsettings_get("org.gnome.shell", "enabled-extensions").unwrap_or_default();
    let inner = raw.trim().trim_start_matches('[').trim_end_matches(']');
    if inner.is_empty() {
        return Vec::new();
    }
    inner
        .split(',')
        .map(|s| s.trim().trim_matches('\'').trim_matches('"').to_string())
        .filter(|s| !s.is_empty())
        .collect()
}

fn set_enabled_extensions(exts: &[String]) -> Result<(), String> {
    let formatted: Vec<String> = exts.iter().map(|e| format!("'{}'", e)).collect();
    let value = format!("[{}]", formatted.join(", "));
    gsettings_set("org.gnome.shell", "enabled-extensions", &value)
}

fn set_extension_state(uuid: &str, enable: bool) -> Result<(), String> {
    let mut exts = get_enabled_extensions();
    if enable {
        if !exts.contains(&uuid.to_string()) {
            exts.push(uuid.to_string());
        }
    } else {
        exts.retain(|e| e != uuid);
    }
    set_enabled_extensions(&exts)
}

// ── Desktop Effects ──

pub fn get_effects_state() -> Result<bool, String> {
    let exts = get_enabled_extensions();
    Ok(exts.contains(&LIQUID_GLASS_UUID.to_string()))
}

pub fn set_effects(use_liquid_glass: bool) -> Result<(), String> {
    if use_liquid_glass {
        set_extension_state(BLUR_MY_SHELL_UUID, false)?;
        set_extension_state(LIQUID_GLASS_UUID, true)?;
        apply_glass_settings()?;
    } else {
        set_extension_state(LIQUID_GLASS_UUID, false)?;
        set_extension_state(BLUR_MY_SHELL_UUID, true)?;
        apply_blur_settings()?;
    }
    Ok(())
}

fn apply_blur_settings() -> Result<(), String> {
    let schema = "org.gnome.shell.extensions.dash-to-dock";
    gsettings_set(schema, "background-opacity", "0.8")?;
    gsettings_set(schema, "custom-theme-shrink", "false")?;
    gsettings_set(schema, "show-show-apps-button", "false")?;
    gsettings_set(schema, "height-fraction", "0.9")?;
    gsettings_set(schema, "apply-custom-theme", "true")?;
    gsettings_set(schema, "transparency-mode", "'FIXED'")?;
    gsettings_set(schema, "customize-alphas", "false")?;
    Ok(())
}

fn apply_glass_settings() -> Result<(), String> {
    let dock = "org.gnome.shell.extensions.dash-to-dock";
    gsettings_set(dock, "background-opacity", "0.0")?;
    gsettings_set(dock, "custom-theme-shrink", "false")?;
    gsettings_set(dock, "show-show-apps-button", "false")?;
    gsettings_set(dock, "height-fraction", "0.9")?;
    gsettings_set(dock, "apply-custom-theme", "false")?;
    gsettings_set(dock, "transparency-mode", "'FIXED'")?;
    gsettings_set(dock, "customize-alphas", "true")?;
    gsettings_set(dock, "min-alpha", "0.0")?;
    gsettings_set(dock, "max-alpha", "0.0")?;

    let glass = "org.gnome.shell.extensions.liquid-glass";
    gsettings_set(glass, "application-blur-radius", "9")?;
    gsettings_set(glass, "application-content-opacity", "1.0")?;
    gsettings_set(glass, "application-corner-radius", "17.0")?;
    gsettings_set(glass, "application-glass-all-windows", "false")?;
    gsettings_set(glass, "application-tint-color", "'#000000'")?;
    gsettings_set(glass, "application-tint-strength", "0.06")?;
    gsettings_set(glass, "application-window-whitelist", "[]")?;
    gsettings_set(glass, "dock-corner-radius", "24.0")?;
    gsettings_set(glass, "dock-glass-expand", "3")?;
    gsettings_set(glass, "dock-tint-color", "'#000000'")?;
    gsettings_set(glass, "enable-application-glass", "false")?;
    gsettings_set(glass, "enable-menu-glass", "true")?;
    gsettings_set(glass, "enable-quick-settings-glass", "false")?;
    gsettings_set(glass, "menu-tint-color", "'#000000'")?;
    gsettings_set(glass, "notification-tint-color", "'#000000'")?;
    gsettings_set(glass, "osd-tint-color", "'#000000'")?;
    gsettings_set(glass, "output-logs", "false")?;
    Ok(())
}

// ── Cleanup ──

pub fn run_cleanup() -> Result<(), String> {
    let scripts = [
        vec!["sudo", "/usr/libexec/pulsar-cleanup-live.sh"],
        vec!["pkexec", "/usr/libexec/pulsar-cleanup-live.sh"],
    ];
    for cmd in &scripts {
        let out = Command::new(cmd[0])
            .args(&cmd[1..])
            .output();
        if let Ok(o) = out {
            if o.status.success() {
                return Ok(());
            }
        }
    }
    Err("Could not run cleanup with sudo or pkexec".into())
}
